contracts/validator/contract/XDCValidatorV2.sol

```diff
@@ -0,0 +1,518 @@
+        msg.sender.transfer(cap);
+        emit Withdraw(msg.sender, _blockNumber, cap);
+    }
+
+    function removeZeroAddresses(
+        address[] memory addresses
+    ) private pure returns (address[] memory) {
+        address[] memory newAddresses = new address[](addresses.length);
+        uint256 j = 0;
+        for (uint256 i = 0; i < addresses.length; i++) {
+            if (addresses[i] != address(0)) {
+                newAddresses[j] = addresses[i];
+                j++;
+            }
+        }
+        // Resize the array.
+        assembly {
+            mstore(newAddresses, j)
+        }
+        return newAddresses;
+    }
+
+    function removeCandidatesZeroAddresses() external {
+        address[] memory newAddresses = new address[](candidates.length);
+        uint256 j = 0;
+        for (uint256 i = 0; i < candidates.length; i++) {
+            if (candidates[i] != address(0)) {
+                newAddresses[j] = candidates[i];
+                j++;
+            }
+        }
+        // Resize the array.
+        assembly {
+            mstore(newAddresses, j)
+        }
+        candidates = newAddresses;
+    }
+
+    function removeOwnersZeroAddresses() external {
+        address[] memory newAddresses = new address[](owners.length);
+        uint256 j = 0;
+        for (uint256 i = 0; i < owners.length; i++) {
+            if (owners[i] != address(0)) {
+                newAddresses[j] = owners[i];
+                j++;
+            }
+        }
+        // Resize the array.
+        assembly {
+            mstore(newAddresses, j)
+        }
+        owners = newAddresses;
+    }
+
+    // Efficiently remove _candidate from the candidates array
+    function deleteCandidate(address candidate) private {
+        uint256 candidatesLength = candidates.length;
+        for (uint256 i = 0; i < candidatesLength; i++) {
+            if (candidates[i] == candidate) {
+                candidates[i] = candidates[candidatesLength - 1];
+                delete candidates[candidatesLength - 1];
+                candidates.length--; // Manually decrease the array length
+                break;
+            }
+        }
+    }
+
+    // Efficiently remove the invalid owner from the owners array
+    function deleteOwner(address owner) private {
+        uint256 ownersLength = owners.length;
+        for (uint k = 0; k < ownersLength; k++) {
+            if (owners[k] == owner) {
+                owners[k] = owners[ownersLength - 1]; // Swap with the last element
+                delete owners[ownersLength - 1]; // Delete the last element
+                owners.length--; // Decrease the array size
+                ownerCount--; // Decrease the owner count
+                break;
+            }
+        }
+    }
+
+    // isOwner : check if the given address is an owner or not.
+    function isOwner(address owner) public view returns (bool) {
+        for (uint i = 0; i < owners.length; i++) {
```

Comment:
gas will be saved if cache storage variable owners before loop

Reply:
just for query , not need care about gas